crypto/pkcs7/pkcs7.go

```diff
@@ -0,0 +1,122 @@
+//
+//			ContentInfo ::= SEQUENCE {
+//				contentType ContentType,
+//				content [0] EXPLICIT ANY DEFINED BY contentType OPTIONAL
+//			}
+//
+// There are 6 possible ContentTypes, data, signedData, envelopedData,
+// signedAndEnvelopedData, digestedData, and encryptedData.  Here onlysignedData is
+// implemented, as the degenerate case of signedData without a signature is the typical
+// format for transferring certificates and CRLS. The ContentType signedData has the
+// form:
+//
+//			signedData ::= SEQUENCE {
+//				version Version,
+//				digestAlgorithms DigestAlgorithmIdentifiers,
+//				contentInfo ContentInfo,
+//				certificates [0] IMPLICIT ExtendedCertificatesAndCertificates OPTIONAL
+//				crls [1] IMPLICIT CertificateRevocationLists OPTIONAL,
+//				signerInfos SignerInfos
+//			}
+//
+// 	As of yet signerInfos and digestAlgorithms are not parsed, as they are not relevant to
+//  this system's use of PKCS #7 data.  Version is an integer type, note that PKCS #7 is
+//	recursive, this second layer of ContentInfo is similar ignored for our degenerate
+//  usage.  The ExtendedCertificatesAndCertificates type consists of a sequence of choices
+//  between PKCS #6 extended certificates andx509 certificates.  Any sequence consisting
+//	of any number of  extended certificates is not yet supported in this implementation
+
+package pkcs7
+
+import (
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/asn1"
+	"errors"
+)
+
+// Types used for asn1 Unmarshalling
+
+type signedData struct {
+	Version          int
+	DigestAlgorithms asn1.RawValue
+	ContentInfo      asn1.RawValue
+	Certificates     asn1.RawValue `asn1:"optional" asn1:"tag:0"`
+	Crls             asn1.RawValue `asn1:optional`
+	SignerInfos      asn1.RawValue
+}
+
+type content struct {
+	SignedData signedData
+}
+
+type initPKCS7 struct {
+	Raw         asn1.RawContent
+	ContentType asn1.ObjectIdentifier
+	Content     content `asn1:"tag:0"`
+}
+
+// External representation of the PKCS7 structure
+
+type PKCS7 struct {
+	Raw          asn1.RawContent
+	Version      int
+	Certificates []*x509.Certificate
+	Crl          *pkix.CertificateList
+}
+
+func ParsePKCS7(raw []byte) (pkcs7 *PKCS7, err error) {
+
+	var init_pkcs7 initPKCS7
+	_, err = asn1.Unmarshal(raw, &init_pkcs7)
+	if err != nil {
+		return nil, err
+	}
+
+	pkcs7 = new(PKCS7)
+	pkcs7.Raw = init_pkcs7.Raw
+	pkcs7.Version = init_pkcs7.Content.SignedData.Version
+
+	if len(init_pkcs7.Content.SignedData.Certificates.Bytes) == 0 {
+		pkcs7.Certificates = nil
+	} else {
+		pkcs7.Certificates, err = x509.ParseCertificates(init_pkcs7.Content.SignedData.Certificates.Bytes)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if len(init_pkcs7.Content.SignedData.Crls.Bytes) == 0 {
+		pkcs7.Crl = nil
+	} else {
+		pkcs7.Crl, err = x509.ParseDERCRL(init_pkcs7.Content.SignedData.Crls.Bytes)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return pkcs7, nil
+
+}
+
+// Retrieves the certificates from a PKCS7 structure
```

Comment:
The `Certificates` field is exported, so I'm not sure that this function is needed.